hazelcast/src/main/java/com/hazelcast/spi/impl/InternalCompletableFuture.java

```diff
@@ -77,63 +78,125 @@ public V joinInternal() {
         return super.thenApplyAsync(fn, defaultExecutor());
     }
 
+    @Override
+    public <U> CompletableFuture<U> thenApplyAsync(Function<? super V, ? extends U> fn) {
+        return super.thenApplyAsync(fn, defaultExecutor());
+    }
+
     @Override
     public CompletableFuture<Void> thenAccept(Consumer<? super V> action) {
         return super.thenAcceptAsync(action, defaultExecutor());
     }
 
+    @Override
+    public CompletableFuture<Void> thenAcceptAsync(Consumer<? super V> action) {
+        return super.thenAcceptAsync(action, defaultExecutor());
+    }
+
     @Override
     public CompletableFuture<Void> thenRun(Runnable action) {
         return super.thenRunAsync(action, defaultExecutor());
     }
 
+    @Override
+    public CompletableFuture<Void> thenRunAsync(Runnable action) {
+        return super.thenRunAsync(action, defaultExecutor());
+    }
+
     @Override
     public <U, V1> CompletableFuture<V1> thenCombine(CompletionStage<? extends U> other,
                                                      BiFunction<? super V, ? super U, ? extends V1> fn) {
         return super.thenCombineAsync(other, fn, defaultExecutor());
     }
 
+    @Override
+    public <U, V1> CompletableFuture<V1> thenCombineAsync(CompletionStage<? extends U> other,
+                                                          BiFunction<? super V, ? super U, ? extends V1> fn) {
+        return super.thenCombineAsync(other, fn,  defaultExecutor());
+    }
+
     @Override
     public <U> CompletableFuture<Void> thenAcceptBoth(CompletionStage<? extends U> other,
                                                       BiConsumer<? super V, ? super U> action) {
         return super.thenAcceptBothAsync(other, action, defaultExecutor());
     }
 
+    @Override
+    public <U> CompletableFuture<Void> thenAcceptBothAsync(CompletionStage<? extends U> other,
+                                                           BiConsumer<? super V, ? super U> action) {
+        return super.thenAcceptBothAsync(other, action, defaultExecutor());
+    }
+
     @Override
     public CompletableFuture<Void> runAfterBoth(CompletionStage<?> other, Runnable action) {
         return super.runAfterBothAsync(other, action, defaultExecutor());
     }
 
+    @Override
+    public CompletableFuture<Void> runAfterBothAsync(CompletionStage<?> other, Runnable action) {
+        return super.runAfterBothAsync(other, action, defaultExecutor());
+    }
+
     @Override
     public <U> CompletableFuture<U> applyToEither(CompletionStage<? extends V> other, Function<? super V, U> fn) {
         return super.applyToEitherAsync(other, fn, defaultExecutor());
     }
 
+    @Override
+    public <U> CompletableFuture<U> applyToEitherAsync(CompletionStage<? extends V> other, Function<? super V, U> fn) {
```

Comment:
@vbekiaris might verify this but as far as I understand, we choose to change the `defaultExecutor` on purpose for even sync ones because using the completion thread might lead to deadlocks in Hazelcast. 
The behaviour change is allowed because on `CompletionStage` class level javadoc, you can see the following:
`The execution properties of default and async modes are specified by CompletionStage implementations, not this interface.`

Reply:
This is exactly the case. This behaviour is documented in the javadoc of data structures which provide an async API eg [`IMap`](https://github.com/hazelcast/hazelcast/blob/40d555dc1110da9f63fd66f504c83d4382d0384b/hazelcast/src/main/java/com/hazelcast/map/IMap.java#L231-L243)

Reply:
So, practically the two APIs are the same methods for out implementation and it just a duplicate API for us.